### PR TITLE
add O2 option in gpt eval

### PR DESCRIPTION
**Pull request by blahBlahhhJ**
# What does this PR do ?

Add an option to run amp O2 in gpt_eval/inference

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/megatron_gpt_eval.py
Comment (maanug-nv) [inline examples/nlp/language_modeling/megatron_gpt_eval.py:1]: I think we can enable this for fp16 as well. So something like:
```
if trainer.precision in [16 '16', '16-mixed', 'bf16', 'bf16-mixed'] and cfg.get('megatron_amp_O2', False):
    pretrained_cfg.megatron_amp_O2 = True
else:
    pretrained_cfg.megatron_amp_O2 = False
```

File: examples/nlp/language_modeling/megatron_gpt_eval.py [hunk 1]
```diff
@@ -213,6 +213,8 @@ def main(cfg) -> None:
             pretrained_cfg.precision = trainer.precision
             if trainer.precision == "16":
                 pretrained_cfg.megatron_amp_O2 = False
+            elif trainer.precision in ['bf16', 'bf16-mixed'] and cfg.get('megatron_amp_O2', False):
```
Comment (maanug-nv) [inline examples/nlp/language_modeling/megatron_gpt_eval.py:216]: I think we can enable this for fp16 as well. So something like:
```
if trainer.precision in [16 '16', '16-mixed', 'bf16', 'bf16-mixed'] and cfg.get('megatron_amp_O2', False):
    pretrained_cfg.megatron_amp_O2 = True
else:
    pretrained_cfg.megatron_amp_O2 = False
```

Comment (blahBlahhhJ) [reply examples/nlp/language_modeling/megatron_gpt_eval.py:216]: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/language_modeling/megatron_base_model.py#L767
fp16 not supported in amp O2.

File: examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml [hunk 2]
```diff
@@ -21,6 +21,7 @@ trainer:
 tensor_model_parallel_size: -1
 pipeline_model_parallel_size: -1
 pipeline_model_parallel_split_rank: -1 # used for encoder and decoder model (0 for others)
+megatron_amp_O2: False
```
Comment (ericharper) [inline examples/nlp/language_modeling/conf/megatron_gpt_inference.yaml:24]: Could you add a comment here explaining what this does for inference?

File: examples/nlp/language_modeling/conf/megatron_llama_inference.yaml [hunk 3]
```diff
@@ -21,6 +21,7 @@ trainer:
 tensor_model_parallel_size: -1
 pipeline_model_parallel_size: -1
 pipeline_model_parallel_split_rank: -1 # used for encoder and decoder model (0 for others)
+megatron_amp_O2: False
```
Comment (ericharper) [inline examples/nlp/language_modeling/conf/megatron_llama_inference.yaml:24]: Could you add a comment here explaining what this does for inference?

Comment (blahBlahhhJ) [reply examples/nlp/language_modeling/conf/megatron_llama_inference.yaml:24]: done

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**